cpp/src_prims/distance/pairwise_distance_base.cuh

```diff
@@ -0,0 +1,182 @@
+ * @tparam bool Sqrt if the square root is computed or not
+ * @tparam Policy struct which tunes the Contraction kernel
+ * @tparam CoreLambda, tells how to accumulate an x and y into 
+    acc. its signature:
+    template <typename AccT, typename DataT> void core_lambda(AccT& acc,
+      const DataT& x, const DataT& y)
+ * @tparam EpilogueLambda applies an elementwise function to compute final 
+    values. Its signature is:
+    template <typename AccT, typename DataT> void epilogue_lambda
+    (AccT acc[][], DataT* sxNorm, DataT* sxNorm);
+ * @tparam FinalLambda the final lambda called on final distance value
+ * @param[in] x input matrix
+ * @param[in] y input matrix
+ * @param[in] m number of rows of A and C/D
+ * @param[in] n number of columns of B and C/D
+ * @param[in] k number of cols of A and rows of B
+ * @param[in] xn row norms of input matrix A. Required for expanded L2
+ * @param[in] yn row norms of input matrix B. Required for expanded L2
+ * @param[output] pD output matrix
+ * @param[in] smem shared mem buffer for intermediate storage of A, B, xn & yn.
+ * @param core_op the core accumulation operation lambda
+ * @param epilog_op the epilog operation lambda
+ * @param fin_op the final gemm epilogue lambda
+ */
+template <typename DataT, typename AccT, typename OutT, typename IdxT,
+          bool Sqrt, typename Policy, typename CoreLambda,
+          typename EpilogueLambda, typename FinalLambda,
+          typename BaseClass =
+            raft::linalg::Contractions_NT<DataT, IdxT, Policy>>
+struct PairwiseDistances : public BaseClass {
+ private:
+  typedef Policy P;
+  const DataT* xn;
+  const DataT* yn;
+  DataT* sxNorm;
+  DataT* syNorm;
+  OutT* dOutput;
+  char* smem;
+  CoreLambda core_op;
+  EpilogueLambda epilog_op;
+  FinalLambda fin_op;
+
+  AccT acc[P::AccRowsPerTh][P::AccColsPerTh];
+
+ public:
+  // Constructor for euclidean expanded
+  DI PairwiseDistances(const DataT* _x, const DataT* _y, IdxT _m, IdxT _n,
+                       IdxT _k, const DataT* _xn, const DataT* _yn,
+                       OutT* _dOutput, char* _smem, CoreLambda _core_op,
+                       EpilogueLambda _epilog_op, FinalLambda _fin_op)
+    : BaseClass(_x, _y, _m, _n, _k, _smem),
+      sxNorm((DataT*)_smem),
+      syNorm(&(sxNorm[P::Mblk])),
+      xn(_xn),
+      yn(_yn),
+      dOutput(_dOutput),
+      smem(_smem),
+      core_op(_core_op),
+      epilog_op(_epilog_op),
+      fin_op(_fin_op) {}
+
+  // Constructor for euclidean unexpanded
+  DI PairwiseDistances(const DataT* _x, const DataT* _y, IdxT _m, IdxT _n,
+                       IdxT _k, OutT* _dOutput, char* _smem,
+                       CoreLambda _core_op, EpilogueLambda _epilog_op,
+                       FinalLambda _fin_op)
+    : BaseClass(_x, _y, _m, _n, _k, _smem),
+      dOutput(_dOutput),
+      smem(_smem),
+      core_op(_core_op),
+      epilog_op(_epilog_op),
+      fin_op(_fin_op) {}
+
+  DI void run() {
+    prolog();
+    loop();
+    epilog();
```

Comment:
In case for some use-cases we want to endup reusing the shared-memory that we are using to store XY in the `loop`, then to avoid any hazards, it is safer to put a `__syncthreads()` before this line.

Reply:
as not all distance metrics require reuse of shared memory for sx/syNorm so it'll be unnecessary `__syncthreads()` issued for them. 
I think we would need to rely on user to add `__syncthreads()` to avoid this redundant call for such cases.

Reply:
Good point! Agree with you.